File: packages/nightgauge-vscode/tests/contracts/contract.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "fs";
import { resolve } from "path";

const FIXTURES_DIR = resolve(__dirname, "../../../../tests/fixtures/contracts");

function loadFixture(name: string): Record<string, unknown> {
  return JSON.parse(readFileSync(resolve(FIXTURES_DIR, name), "utf-8"));
}

// ─── Field declarations ────────────────────────────────────────────────────
// Each set of fields must be kept in sync with the corresponding interface
// in src/services/IpcClient.ts. When Go adds a field, the fixture is updated,
// and the "no unknown fields" test below will fail until the field is added
// to either TS_FIELDS or GO_ONLY_FIELDS.

const BOARD_ITEM_TS_FIELDS: Record<string, string> = {
  id: "string",
  number: "number",
  title: "string",
  state: "string",
  status: "string",
  priority: "string",
  size: "string",
  labels: "object", // array
  repo: "string",
  url: "string",
  isEpic: "boolean",
  blockedBy: "object", // array
  blocking: "object", // array
  subIssues: "object", // array
  parentIssueNumber: "number",
  parentIssueTitle: "string",
};

// Fields Go serializes but TS BoardItem does not declare.
// When TS adopts a field, move it from here to BOARD_ITEM_TS_FIELDS.
const BOARD_ITEM_GO_ONLY_FIELDS = new Set([
  "nodeId",
  "pipelineStage",
  "createdAt",
  "updatedAt",
  "isPR",
]);

const BOARD_ITEM_ALL_KNOWN = new Set([
  ...Object.keys(BOARD_ITEM_TS_FIELDS),
  ...BOARD_ITEM_GO_ONLY_FIELDS,
]);

const ISSUE_TS_FIELDS: Record<string, string> = {
  number: "number",
  title: "string",
  body: "string",
  state: "string",
  labels: "object", // array
  assignees: "object", // array
  url: "string",
  blockedBy: "object", // array
  blocking: "object", // array
  subIssues: "object", // array
};

// Optional TS fields that may or may not be in the fixture
const ISSUE_TS_OPTIONAL: Record<string, string> = {
  id: "string",
  stateReason: "string",
  isEpic: "boolean",
  parentIssueId: "string",
  parentIssueNumber: "number",
  milestone: "string",
};

// Fields Go serializes but TS IssueDetail does not declare.
const ISSUE_GO_ONLY_FIELDS = new Set(["nodeId", "repo"]);

const ISSUE_ALL_KNOWN = new Set([
  ...Object.keys(ISSUE_TS_FIELDS),
  ...Object.keys(ISSUE_TS_OPTIONAL),
  ...ISSUE_GO_ONLY_FIELDS,
]);

// ─── Nested ref schemas ────────────────────────────────────────────────────

const BLOCKING_REF_FIELDS: Record<string, string> = {
  number: "number",
  title: "string",
  state: "string",
  repo: "string",
};

const BLOCKING_REF_GO_ONLY = new Set(["nodeId"]);
const BLOCKING_REF_ALL_KNOWN = new Set([
  ...Object.keys(BLOCKING_REF_FIELDS),
  ...BLOCKING_REF_GO_ONLY,
]);

const SUB_ISSUE_REF_FIELDS: Record<string, string> = {
  number: "number",
  title: "string",
  state: "string",
  repo: "string",
  labels: "object", // array
};

const SUB_ISSUE_REF_GO_ONLY = new Set(["nodeId"]);
const SUB_ISSUE_REF_ALL_KNOWN = new Set([
  ...Object.keys(SUB_ISSUE_REF_FIELDS),
  ...SUB_ISSUE_REF_GO_ONLY,
]);

// ─── Helper ────────────────────────────────────────────────────────────────

function validateFields(
  obj: Record<string, unknown>,
  requiredFields: Record<string, string>,
  allKnown: Set<string>,
  label: string
) {
  // Every required TS field must exist with the expected type
  for (const [field, expectedType] of Object.entries(requiredFields)) {
    expect(obj, `${label}: missing required field "${field}"`).toHaveProperty(field);
    expect(typeof obj[field], `${label}: "${field}" type mismatch`).toBe(expectedType);
  }

  // Every fixture key must be in the known set (catches Go additions)
  for (const key of Object.keys(obj)) {
    expect(
      allKnown.has(key),
      `${label}: unknown field "${key}" in fixture — add to TS interface or GO_ONLY_FIELDS`
    ).toBe(true);
  }
}

function validateRefArray(
  arr: unknown,
  refFields: Record<string, string>,
  allKnown: Set<string>,
  label: string
) {
  expect(Array.isArray(arr), `${label}: expected array`).toBe(true);
  for (const [i, item] of (arr as unknown[]).entries()) {
    validateFields(item as Record<string, unknown>, refFields, allKnown, `${label}[${i}]`);
  }
}

// ─── BoardItem contract ────────────────────────────────────────────────────

describe("BoardItem contract (Go → TypeScript)", () => {
  const fixture = loadFixture("board-item.json");

  it("contains all required TS fields with correct types", () => {
    validateFields(fixture, BOARD_ITEM_TS_FIELDS, BOARD_ITEM_ALL_KNOWN, "BoardItem");
  });

  it("has no unknown top-level fields", () => {
    for (const key of Object.keys(fixture)) {
      expect(
        BOARD_ITEM_ALL_KNOWN.has(key),
        `BoardItem: unknown field "${key}" — add to TS interface or GO_ONLY_FIELDS`
      ).toBe(true);
    }
  });

  it("subIssues entries match SubIssueRef schema", () => {
    validateRefArray(
      fixture.subIssues,
      SUB_ISSUE_REF_FIELDS,
      SUB_ISSUE_REF_ALL_KNOWN,
      "BoardItem.subIssues"
    );
  });

  it("blockedBy entries match BlockingRef schema", () => {
    validateRefArray(
      fixture.blockedBy,
      BLOCKING_REF_FIELDS,
      BLOCKING_REF_ALL_KNOWN,
      "BoardItem.blockedBy"
    );
  });

  it("blocking entries match BlockingRef schema", () => {
    validateRefArray(
      fixture.blocking,
      BLOCKING_REF_FIELDS,
      BLOCKING_REF_ALL_KNOWN,
      "BoardItem.blocking"
    );
  });
});

// ─── Issue contract ────────────────────────────────────────────────────────

describe("Issue contract (Go → TypeScript)", () => {
  const fixture = loadFixture("issue.json");

  it("contains all required TS fields with correct types", () => {
    validateFields(fixture, ISSUE_TS_FIELDS, ISSUE_ALL_KNOWN, "IssueDetail");
  });

  it("optional TS fields have correct types when present", () => {
    for (const [field, expectedType] of Object.entries(ISSUE_TS_OPTIONAL)) {
      if (field in fixture) {
        expect(typeof fixture[field], `IssueDetail: optional field "${field}" type mismatch`).toBe(
          expectedType
        );
      }
    }
  });

  it("has no unknown top-level fields", () => {
    for (const key of Object.keys(fixture)) {
      expect(
        ISSUE_ALL_KNOWN.has(key),
        `IssueDetail: unknown field "${key}" — add to TS interface or GO_ONLY_FIELDS`
      ).toBe(true);
    }
  });

  it("subIssues entries match SubIssueRef schema", () => {
    validateRefArray(
      fixture.subIssues,
      SUB_ISSUE_REF_FIELDS,
      SUB_ISSUE_REF_ALL_KNOWN,
      "IssueDetail.subIssues"
    );
  });

  it("blockedBy entries match BlockingRef schema", () => {
    validateRefArray(
      fixture.blockedBy,
      BLOCKING_REF_FIELDS,
      BLOCKING_REF_ALL_KNOWN,
      "IssueDetail.blockedBy"
    );
  });

  it("blocking entries match BlockingRef schema", () => {
    validateRefArray(
      fixture.blocking,
      BLOCKING_REF_FIELDS,
      BLOCKING_REF_ALL_KNOWN,
      "IssueDetail.blocking"
    );
  });
});

File: pkg/types/contract_test.go
package types_test

import (
	"encoding/json"
	"os"
	"reflect"
	"sort"
	"strings"
	"testing"
	"time"

	"github.com/nightgauge/nightgauge/pkg/types"
)

const fixturesDir = "../../tests/fixtures/contracts"

// structJSONKeys extracts all JSON field names from a struct type via reflection.
// It returns sorted keys so comparisons are deterministic.
func structJSONKeys(v interface{}) []string {
	t := reflect.TypeOf(v)
	var keys []string
	for i := 0; i < t.NumField(); i++ {
		tag := t.Field(i).Tag.Get("json")
		if tag == "" || tag == "-" {
			continue
		}
		keys = append(keys, strings.Split(tag, ",")[0])
	}
	sort.Strings(keys)
	return keys
}

// fixtureKeys loads a JSON fixture file and returns its sorted top-level keys.
func fixtureKeys(t *testing.T, path string) []string {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read fixture %s: %v", path, err)
	}
	var m map[string]interface{}
	if err := json.Unmarshal(data, &m); err != nil {
		t.Fatalf("failed to parse fixture %s: %v", path, err)
	}
	var keys []string
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// loadFixture reads a JSON fixture and unmarshals it into a generic map.
func loadFixture(t *testing.T, path string) map[string]interface{} {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read fixture %s: %v", path, err)
	}
	var m map[string]interface{}
	if err := json.Unmarshal(data, &m); err != nil {
		t.Fatalf("failed to parse fixture %s: %v", path, err)
	}
	return m
}

// --- BoardItem contract tests ---

func TestBoardItemContractKeys(t *testing.T) {
	structKeys := structJSONKeys(types.BoardItem{})
	fixtKeys := fixtureKeys(t, fixturesDir+"/board-item.json")

	if !reflect.DeepEqual(structKeys, fixtKeys) {
		missing := diff(structKeys, fixtKeys)
		extra := diff(fixtKeys, structKeys)
		if len(missing) > 0 {
			t.Errorf("BoardItem struct has keys not in fixture: %v — update the fixture", missing)
		}
		if len(extra) > 0 {
			t.Errorf("fixture has keys not in BoardItem struct: %v — update the struct or fixture", extra)
		}
	}
}

func TestBoardItemContractValues(t *testing.T) {
	item := types.BoardItem{
		ID:            "PVI_item123",
		NodeID:        "I_node456",
		Number:        42,
		Title:         "Add photo upload feature",
		State:         "OPEN",
		Status:        "Ready",
		Priority:      types.PriorityP1,
		Size:          types.SizeM,
		PipelineStage: "feature-dev",
		Labels:        []string{"type:feature", "priority:high"},
		Repo:          "nightgauge/nightgauge",
		URL:           "https://github.com/nightgauge/nightgauge/issues/42",
		CreatedAt:     time.Date(2025, 1, 15, 10, 30, 0, 0, time.UTC),
		UpdatedAt:     time.Date(2025, 1, 16, 14, 45, 0, 0, time.UTC),
		IsPR:          false,
		IsEpic:        false,
		ParentNumber:  10,
		ParentTitle:   "Epic: Photo Management",
		SubIssues: []types.SubIssueRef{
			{NodeID: "I_sub789", Number: 43, Title: "Implement upload API", State: "OPEN", Repo: "nightgauge/nightgauge", Labels: []string{"size:M"}},
		},
		BlockedBy: []types.BlockingRef{
			{NodeID: "I_block101", Number: 40, Title: "Setup storage backend", State: "CLOSED", Repo: "nightgauge/nightgauge"},
		},
		Blocking: []types.BlockingRef{
			{NodeID: "I_block202", Number: 44, Title: "Add photo gallery view", State: "OPEN", Repo: "nightgauge/nightgauge"},
		},
	}

	got, err := json.Marshal(item)
	if err != nil {
		t.Fatalf("failed to marshal BoardItem: %v", err)
	}

	want, err := os.ReadFile(fixturesDir + "/board-item.json")
	if err != nil {
		t.Fatalf("failed to read fixture: %v", err)
	}

	var gotMap, wantMap map[string]interface{}
	if err := json.Unmarshal(got, &gotMap); err != nil {
		t.Fatalf("failed to parse marshaled JSON: %v", err)
	}
	if err := json.Unmarshal(want, &wantMap); err != nil {
		t.Fatalf("failed to parse fixture JSON: %v", err)
	}

	if !reflect.DeepEqual(gotMap, wantMap) {
		t.Errorf("BoardItem serialization does not match fixture\nGot:  %s\nWant: %s", got, want)
	}
}

// --- Issue contract tests ---

func TestIssueContractKeys(t *testing.T) {
	structKeys := structJSONKeys(types.Issue{})
	fixtKeys := fixtureKeys(t, fixturesDir+"/issue.json")

	if !reflect.DeepEqual(structKeys, fixtKeys) {
		missing := diff(structKeys, fixtKeys)
		extra := diff(fixtKeys, structKeys)
		if len(missing) > 0 {
			t.Errorf("Issue struct has keys not in fixture: %v — update the fixture", missing)
		}
		if len(extra) > 0 {
			t.Errorf("fixture has keys not in Issue struct: %v — update the struct or fixture", extra)
		}
	}
}

func TestIssueContractValues(t *testing.T) {
	issue := types.Issue{
		NodeID:            "I_node456",
		Number:            42,
		Title:             "Add photo upload feature",
		Body:              "## Summary\nImplement photo upload with drag-and-drop support.",
		State:             "OPEN",
		StateReason:       "REOPENED",
		Labels:            []string{"type:feature", "priority:high"},
		Repo:              "nightgauge/nightgauge",
		URL:               "https://github.com/nightgauge/nightgauge/issues/42",
		Assignees:         []string{"octocat"},
		IsEpic:            true,
		Milestone:         "Sprint 1",
		ParentIssueID:     "I_parent100",
		ParentIssueNumber: 40,
		SubIssues: []types.SubIssueRef{
			{NodeID: "I_sub789", Number: 43, Title: "Implement upload API", State: "OPEN", Repo: "nightgauge/nightgauge", Labels: []string{"size:M"}},
		},
		BlockedBy: []types.BlockingRef{
			{NodeID: "I_block101", Number: 40, Title: "Setup storage backend", State: "CLOSED", Repo: "nightgauge/nightgauge"},
		},
		Blocking: []types.BlockingRef{
			{NodeID: "I_block202", Number: 44, Title: "Add photo gallery view", State: "OPEN", Repo: "nightgauge/nightgauge"},
		},
	}

	got, err := json.Marshal(issue)
	if err != nil {
		t.Fatalf("failed to marshal Issue: %v", err)
	}

	want, err := os.ReadFile(fixturesDir + "/issue.json")
	if err != nil {
		t.Fatalf("failed to read fixture: %v", err)
	}

	var gotMap, wantMap map[string]interface{}
	if err := json.Unmarshal(got, &gotMap); err != nil {
		t.Fatalf("failed to parse marshaled JSON: %v", err)
	}
	if err := json.Unmarshal(want, &wantMap); err != nil {
		t.Fatalf("failed to parse fixture JSON: %v", err)
	}

	if !reflect.DeepEqual(gotMap, wantMap) {
		t.Errorf("Issue serialization does not match fixture\nGot:  %s\nWant: %s", got, want)
	}
}

// --- SubIssueRef contract test ---

func TestSubIssueRefContractKeys(t *testing.T) {
	structKeys := structJSONKeys(types.SubIssueRef{})
	// SubIssueRef is validated via the nested objects in board-item.json and issue.json.
	// Load a SubIssueRef from the board-item fixture.
	fixture := loadFixture(t, fixturesDir+"/board-item.json")
	subIssues, ok := fixture["subIssues"].([]interface{})
	if !ok || len(subIssues) == 0 {
		t.Fatal("fixture board-item.json must have at least one subIssues entry")
	}
	sub, ok := subIssues[0].(map[string]interface{})
	if !ok {
		t.Fatal("subIssues[0] is not an object")
	}

	var subKeys []string
	for k := range sub {
		subKeys = append(subKeys, k)
	}
	sort.Strings(subKeys)

	if !reflect.DeepEqual(structKeys, subKeys) {
		t.Errorf("SubIssueRef struct keys do not match fixture subIssues[0] keys\nStruct: %v\nFixture: %v", structKeys, subKeys)
	}
}

// --- BlockingRef contract test ---

func TestBlockingRefContractKeys(t *testing.T) {
	structKeys := structJSONKeys(types.BlockingRef{})
	fixture := loadFixture(t, fixturesDir+"/board-item.json")
	blockedBy, ok := fixture["blockedBy"].([]interface{})
	if !ok || len(blockedBy) == 0 {
		t.Fatal("fixture board-item.json must have at least one blockedBy entry")
	}
	block, ok := blockedBy[0].(map[string]interface{})
	if !ok {
		t.Fatal("blockedBy[0] is not an object")
	}

	var blockKeys []string
	for k := range block {
		blockKeys = append(blockKeys, k)
	}
	sort.Strings(blockKeys)

	if !reflect.DeepEqual(structKeys, blockKeys) {
		t.Errorf("BlockingRef struct keys do not match fixture blockedBy[0] keys\nStruct: %v\nFixture: %v", structKeys, blockKeys)
	}
}

// diff returns elements in a that are not in b.
func diff(a, b []string) []string {
	set := make(map[string]bool, len(b))
	for _, s := range b {
		set[s] = true
	}
	var result []string
	for _, s := range a {
		if !set[s] {
			result = append(result, s)
		}
	}
	return result
}

// --- StatusCounts contract tests ---

func TestStatusCountsContractKeys(t *testing.T) {
	structKeys := structJSONKeys(types.StatusCounts{})
	fixtKeys := fixtureKeys(t, fixturesDir+"/status-counts.json")

	if !reflect.DeepEqual(structKeys, fixtKeys) {
		missing := diff(structKeys, fixtKeys)
		extra := diff(fixtKeys, structKeys)
		if len(missing) > 0 {
			t.Errorf("StatusCounts struct has keys not in fixture: %v — update the fixture", missing)
		}
		if len(extra) > 0 {
			t.Errorf("fixture has keys not in StatusCounts struct: %v — update the struct or fixture", extra)
		}
	}
}

func TestStatusCountsContractValues(t *testing.T) {
	sc := types.StatusCounts{
		Ready:      5,
		InProgress: 2,
		InReview:   1,
		Done:       12,
		Backlog:    8,
	}

	got, err := json.Marshal(sc)
	if err != nil {
		t.Fatalf("failed to marshal StatusCounts: %v", err)
	}

	want, err := os.ReadFile(fixturesDir + "/status-counts.json")
	if err != nil {
		t.Fatalf("failed to read fixture: %v", err)
	}

	var gotMap, wantMap map[string]interface{}
	if err := json.Unmarshal(got, &gotMap); err != nil {
		t.Fatalf("failed to parse marshaled JSON: %v", err)
	}
	if err := json.Unmarshal(want, &wantMap); err != nil {
		t.Fatalf("failed to parse fixture JSON: %v", err)
	}

	if !reflect.DeepEqual(gotMap, wantMap) {
		t.Errorf("StatusCounts serialization does not match fixture\nGot:  %s\nWant: %s", got, want)
	}
}

// --- EpicProgress contract tests ---

func TestEpicProgressContractKeys(t *testing.T) {
	structKeys := structJSONKeys(types.EpicProgress{})
	fixtKeys := fixtureKeys(t, fixturesDir+"/epic-progress.json")

	// Remove nested "subIssues" from fixture keys — it's an array, not top-level key mismatch
	if !reflect.DeepEqual(structKeys, fixtKeys) {
		missing := diff(structKeys, fixtKeys)
		extra := diff(fixtKeys, structKeys)
		if len(missing) > 0 {
			t.Errorf("EpicProgress struct has keys not in fixture: %v — update the fixture", missing)
		}
		if len(extra) > 0 {
			t.Errorf("fixture has keys not in EpicProgress struct: %v — update the struct or fixture", extra)
		}
	}
}

func TestEpicProgressContractValues(t *testing.T) {
	ep := types.EpicProgress{
		EpicNodeID: "I_epic001",
		Number:     100,
		Title:      "Epic: User Authentication",
		Repo:       "nightgauge/nightgauge",
		SubIssues: []types.SubIssueRef{
			{NodeID: "I_sub001", Number: 101, Title: "Login page", State: "CLOSED", Repo: "nightgauge/nightgauge"},
			{NodeID: "I_sub002", Number: 102, Title: "Registration page", State: "OPEN", Repo: "nightgauge/nightgauge"},
		},
		Total:           2,
		Closed:          1,
		Open:            1,
		PercentComplete: 50,
	}

	got, err := json.Marshal(ep)
	if err != nil {
		t.Fatalf("failed to marshal EpicProgress: %v", err)
	}

	want, err := os.ReadFile(fixturesDir + "/epic-progress.json")
	if err != nil {
		t.Fatalf("failed to read fixture: %v", err)
	}

	var gotMap, wantMap map[string]interface{}
	if err := json.Unmarshal(got, &gotMap); err != nil {
		t.Fatalf("failed to parse marshaled JSON: %v", err)
	}
	if err := json.Unmarshal(want, &wantMap); err != nil {
		t.Fatalf("failed to parse fixture JSON: %v", err)
	}

	if !reflect.DeepEqual(gotMap, wantMap) {
		t.Errorf("EpicProgress serialization does not match fixture\nGot:  %s\nWant: %s", got, want)
	}
}

// --- PullRequest contract tests ---

func TestPullRequestContractKeys(t *testing.T) {
	structKeys := structJSONKeys(types.PullRequest{})
	fixtKeys := fixtureKeys(t, fixturesDir+"/pull-request.json")

	if !reflect.DeepEqual(structKeys, fixtKeys) {
		missing := diff(structKeys, fixtKeys)
		extra := diff(fixtKeys, structKeys)
		if len(missing) > 0 {
			t.Errorf("PullRequest struct has keys not in fixture: %v — update the fixture", missing)
		}
		if len(extra) > 0 {
			t.Errorf("fixture has keys not in PullRequest struct: %v — update the struct or fixture", extra)
		}
	}
}

func TestPullRequestContractValues(t *testing.T) {
	pr := types.PullRequest{
		NodeID:           "PR_node123",
		Number:           99,
		Title:            "feat: add user auth",
		Body:             "## Summary\n\nImplements user authentication.",
		State:            "OPEN",
		HeadRef:          "feat/99-user-auth",
		BaseRef:          "main",
		Repo:             "nightgauge/nightgauge",
		URL:              "https://github.com/nightgauge/nightgauge/pull/99",
		Mergeable:        "MERGEABLE",
		MergeStateStatus: "CLEAN",
		ReviewStatus:     "APPROVED",
		CheckStatus:      "SUCCESS",
		Labels:           []string{"type:feature"},
		IsDraft:          false,
	}

	got, err := json.Marshal(pr)
	if err != nil {
		t.Fatalf("failed to marshal PullRequest: %v", err)
	}

	want, err := os.ReadFile(fixturesDir + "/pull-request.json")
	if err != nil {
		t.Fatalf("failed to read fixture: %v", err)
	}

	var gotMap, wantMap map[string]interface{}
	if err := json.Unmarshal(got, &gotMap); err != nil {
		t.Fatalf("failed to parse marshaled JSON: %v", err)
	}
	if err := json.Unmarshal(want, &wantMap); err != nil {
		t.Fatalf("failed to parse fixture JSON: %v", err)
	}

	if !reflect.DeepEqual(gotMap, wantMap) {
		t.Errorf("PullRequest serialization does not match fixture\nGot:  %s\nWant: %s", got, want)
	}
}
